### ISSUE TITLE
Every S3 Object is at 0 bytes

### ISSUE THREAD

**Issue by nwmqpa** [pulumi/pulumi#3900]
* Language: Typescript
* Pulumi version: v1.10.1
* @pulumi/aws: 1.21.0
* @pulumi/pulumi: 1.9.1

Recently, I've been unable to upload any S3 object. Each one of them is sized 0 bytes in AWS S3.
Here is the sending code:

```typescript
import * as pulumi from "@pulumi/pulumi";
import * as aws from "@pulumi/aws";
import fs = require('fs');
import path = require('path');
import mime = require('mime');
import command = require('child_process');

// Create an AWS resource (S3 Bucket)

const createBucket = (reactAppName: string): aws.s3.Bucket => {
  return new aws.s3.Bucket(`${reactAppName}-react-webapp`, {
    website: {
      indexDocument: 'index.html'
    },
  });
}

// Reccursivly launch function on folder.
const parseFolder = (folderName: string, func: (filePath: string) => aws.s3.BucketObject): aws.s3.BucketObject[] => {
  return fs.readdirSync(folderName).map(item => {
    const newPath: string = path.join(folderName, item)

    if (fs.statSync(newPath).isDirectory())
      return parseFolder(newPath, func)
    else
      return func(newPath)
  }).flat(Infinity)
}

// Create an S3 Bucket Policy to allow public read of all objects in bucket
// This reusable function can be pulled out into its own module
const publicReadPolicyForBucket = (bucketName: string): string => {
  return JSON.stringify({
    Version: "2012-10-17",
    Statement: [{
      Effect: "Allow",
      Principal: "*",
      Action: [
        "s3:GetObject"
      ],
      Resource: [
        `arn:aws:s3:::${bucketName}/*` // policy refers to bucket name explicitly
      ]
    }]
  })
}


export default (reactAppPath: string, reactAppName: string): {bucket: aws.s3.Bucket, objects: aws.s3.BucketObject[]} => {
  const siteBucket = createBucket(reactAppName);
  const buildDirectory = path.join(process.cwd(), reactAppPath + "/build")
  // Execute a build command
  let objects: aws.s3.BucketObject[] = [];
  command.exec(`cd ${reactAppPath} && npm install && npm run build`,
    (err: Error | null, stdin: string, stdout: string): void => {
      if (err) {
        console.log(stdout);
        console.error(err);
      } else {
        const bucketObjects = parseFolder(buildDirectory, (filePath: string): aws.s3.BucketObject => {
          const relativeFilePath = filePath.replace(buildDirectory + "/", "");
          return new aws.s3.BucketObject(
            relativeFilePath, {
                key: relativeFilePath,
                acl: 'public-read',
                bucket: siteBucket,
                source: new pulumi.asset.FileAsset(filePath),
                contentType: mime.getType(filePath) || undefined,
            }, {
              parent: siteBucket
            }
          )
        })

        new aws.s3.BucketPolicy("bucketPolicy", {
          bucket: siteBucket.bucket,
          policy: siteBucket.bucket.apply(publicReadPolicyForBucket)
        })
        objects = bucketObjects;
      }
    }
  );

  return {
    bucket: siteBucket,
    objects: objects
  }
}
```

All files are present, but none of them is sized greater than 0

Where dit I made it wrong ?

Also, this code works well, althought it's basically the same:
```typescript

// crawlDirectory recursive crawls the provided directory, applying the provided function
// to every file it contains. Doesn't handle cycles from symlinks.
function crawlDirectory(dir: string, f: (_: string) => void) {
    const files = fs.readdirSync(dir);
    for (const file of files) {
        const filePath = `${dir}/${file}`;
        const stat = fs.statSync(filePath);
        if (stat.isDirectory()) {
            crawlDirectory(filePath, f);
        }
        if (stat.isFile()) {
            f(filePath);
        }
    }
}

// Sync the contents of the source directory with the S3 bucket, which will in-turn show up on the CDN.
const webContentsRootPath = path.join(process.cwd(), config.pathToWebsiteContents);
console.log("Syncing contents from local disk at", webContentsRootPath);
crawlDirectory(
    webContentsRootPath,
    (filePath: string) => {
        const relativeFilePath = filePath.replace(webContentsRootPath + "/", "");
        const contentFile = new aws.s3.BucketObject(
            relativeFilePath,
            {
                key: relativeFilePath,

                acl: "public-read",
                bucket: contentBucket,
                contentType: mime.getType(filePath) || undefined,
                source: new pulumi.asset.FileAsset(filePath),
            },
            {
                parent: contentBucket,
            });
    });
````

**Comment by lukehoban** [comment]
I tried for a few minutes to reproduce with your code above - but looks like it depends on some non-trivial external context.

Can you log out what `filePath` is right before calling `new aws.s3.BucketObject`?  

I can't from inspection tell exactly what the salient difference is between the two code examples above - what to you know/intend to be different about these?

**Comment by nwmqpa** [comment]
The second code is from me, while the first is from a co-worker who wanted to modularize my codebase to make some helper functions.

Here's the pulumi output
```
Previewing update (dev):

    pulumi:pulumi:Stack Crm-dev running 
    pulumi:providers:aws east  
    aws:acm:Certificate crm-certificate  
    aws:s3:Bucket crm-react-webapp  
    aws:cloudfront:Distribution s3Distribution  
    aws:route53:Record crmDomain  
@ Previewing update....
    pulumi:pulumi:Stack Crm-dev running /home/tnicollet/Workspace/Nebulis/pulumi-crm/crm/build/asset-manifest.json
    pulumi:pulumi:Stack Crm-dev running /home/tnicollet/Workspace/Nebulis/pulumi-crm/crm/build/favicon.ico
    pulumi:pulumi:Stack Crm-dev running /home/tnicollet/Workspace/Nebulis/pulumi-crm/crm/build/index.html
    pulumi:pulumi:Stack Crm-dev running /home/tnicollet/Workspace/Nebulis/pulumi-crm/crm/build/logo192.png
    pulumi:pulumi:Stack Crm-dev running /home/tnicollet/Workspace/Nebulis/pulumi-crm/crm/build/logo512.png
    pulumi:pulumi:Stack Crm-dev running /home/tnicollet/Workspace/Nebulis/pulumi-crm/crm/build/manifest.json
    pulumi:pulumi:Stack Crm-dev running /home/tnicollet/Workspace/Nebulis/pulumi-crm/crm/build/precache-manifest.3ee4d1c5bcf3b07aaebd61b1a14e081a.js
    pulumi:pulumi:Stack Crm-dev running /home/tnicollet/Workspace/Nebulis/pulumi-crm/crm/build/robots.txt
    pulumi:pulumi:Stack Crm-dev running /home/tnicollet/Workspace/Nebulis/pulumi-crm/crm/build/service-worker.js
    pulumi:pulumi:Stack Crm-dev running /home/tnicollet/Workspace/Nebulis/pulumi-crm/crm/build/static/css/main.d1b05096.chunk.css
    pulumi:pulumi:Stack Crm-dev running /home/tnicollet/Workspace/Nebulis/pulumi-crm/crm/build/static/css/main.d1b05096.chunk.css.map
    pulumi:pulumi:Stack Crm-dev running /home/tnicollet/Workspace/Nebulis/pulumi-crm/crm/build/static/js/2.43075c54.chunk.js
    pulumi:pulumi:Stack Crm-dev running /home/tnicollet/Workspace/Nebulis/pulumi-crm/crm/build/static/js/2.43075c54.chunk.js.LICENSE.txt
    pulumi:pulumi:Stack Crm-dev running /home/tnicollet/Workspace/Nebulis/pulumi-crm/crm/build/static/js/2.43075c54.chunk.js.map
    pulumi:pulumi:Stack Crm-dev running /home/tnicollet/Workspace/Nebulis/pulumi-crm/crm/build/static/js/main.e8a8522c.chunk.js
    pulumi:pulumi:Stack Crm-dev running /home/tnicollet/Workspace/Nebulis/pulumi-crm/crm/build/static/js/main.e8a8522c.chunk.js.map
    pulumi:pulumi:Stack Crm-dev running /home/tnicollet/Workspace/Nebulis/pulumi-crm/crm/build/static/js/runtime-main.2ba79c4d.js
    pulumi:pulumi:Stack Crm-dev running /home/tnicollet/Workspace/Nebulis/pulumi-crm/crm/build/static/js/runtime-main.2ba79c4d.js.map
    aws:s3:BucketObject asset-manifest.json  
    aws:s3:BucketObject favicon.ico  
    aws:s3:BucketObject logo192.png  
    aws:s3:BucketObject static/js/main.e8a8522c.chunk.js  
    aws:s3:BucketObject precache-manifest.3ee4d1c5bcf3b07aaebd61b1a14e081a.js  
    aws:s3:BucketObject static/js/2.43075c54.chunk.js.LICENSE.txt  
    aws:s3:BucketObject static/css/main.d1b05096.chunk.css  
    aws:s3:BucketObject static/js/runtime-main.2ba79c4d.js.map  
    aws:s3:BucketObject index.html  
    aws:s3:BucketObject service-worker.js  
    aws:s3:BucketObject manifest.json  
    aws:s3:BucketObject logo512.png  
    aws:s3:BucketObject robots.txt  
    aws:s3:BucketObject static/js/runtime-main.2ba79c4d.js  
    aws:s3:BucketObject static/css/main.d1b05096.chunk.css.map  
    aws:s3:BucketObject static/js/main.e8a8522c.chunk.js.map  
    aws:s3:BucketPolicy bucketPolicy  
    aws:s3:BucketObject static/js/2.43075c54.chunk.js  
    aws:s3:BucketObject static/js/2.43075c54.chunk.js.map  
    pulumi:pulumi:Stack Crm-dev  18 messages
 
Diagnostics:
  pulumi:pulumi:Stack (Crm-dev):
    /home/tnicollet/Workspace/Nebulis/pulumi-crm/crm/build/asset-manifest.json
    /home/tnicollet/Workspace/Nebulis/pulumi-crm/crm/build/favicon.ico
    /home/tnicollet/Workspace/Nebulis/pulumi-crm/crm/build/index.html
    /home/tnicollet/Workspace/Nebulis/pulumi-crm/crm/build/logo192.png
    /home/tnicollet/Workspace/Nebulis/pulumi-crm/crm/build/logo512.png
    /home/tnicollet/Workspace/Nebulis/pulumi-crm/crm/build/manifest.json
    /home/tnicollet/Workspace/Nebulis/pulumi-crm/crm/build/precache-manifest.3ee4d1c5bcf3b07aaebd61b1a14e081a.js
    /home/tnicollet/Workspace/Nebulis/pulumi-crm/crm/build/robots.txt
    /home/tnicollet/Workspace/Nebulis/pulumi-crm/crm/build/service-worker.js
    /home/tnicollet/Workspace/Nebulis/pulumi-crm/crm/build/static/css/main.d1b05096.chunk.css
    /home/tnicollet/Workspace/Nebulis/pulumi-crm/crm/build/static/css/main.d1b05096.chunk.css.map
    /home/tnicollet/Workspace/Nebulis/pulumi-crm/crm/build/static/js/2.43075c54.chunk.js
    /home/tnicollet/Workspace/Nebulis/pulumi-crm/crm/build/static/js/2.43075c54.chunk.js.LICENSE.txt
    /home/tnicollet/Workspace/Nebulis/pulumi-crm/crm/build/static/js/2.43075c54.chunk.js.map
    /home/tnicollet/Workspace/Nebulis/pulumi-crm/crm/build/static/js/main.e8a8522c.chunk.js
    /home/tnicollet/Workspace/Nebulis/pulumi-crm/crm/build/static/js/main.e8a8522c.chunk.js.map
    /home/tnicollet/Workspace/Nebulis/pulumi-crm/crm/build/static/js/runtime-main.2ba79c4d.js
    /home/tnicollet/Workspace/Nebulis/pulumi-crm/crm/build/static/js/runtime-main.2ba79c4d.js.map
 

Permalink: https://app.pulumi.com/deployment_nebulis/Crm/dev/previews/dafeef7c-ac9e-45ce-862b-0089bec5687c
```

And here is a snapshot of my filesystem

```
> tree /home/tnicollet/Workspace/Nebulis/pulumi-crm/crm/build
drwxr-xr-x    - tnicollet 10 Feb  9:09 /home/tnicollet/Workspace/Nebulis/pulumi-crm/crm/build
.rw-r--r-- 1.0k tnicollet 10 Feb  9:09 ├── asset-manifest.json
.rw-r--r-- 3.1k tnicollet 10 Feb  9:09 ├── favicon.ico
.rw-r--r-- 2.2k tnicollet 10 Feb  9:09 ├── index.html
.rw-r--r-- 5.3k tnicollet 10 Feb  9:09 ├── logo192.png
.rw-r--r-- 9.7k tnicollet 10 Feb  9:09 ├── logo512.png
.rw-r--r--  492 tnicollet 10 Feb  9:09 ├── manifest.json
.rw-r--r--  657 tnicollet 10 Feb  9:09 ├── precache-manifest.3ee4d1c5bcf3b07aaebd61b1a14e081a.js
.rw-r--r--   67 tnicollet 10 Feb  9:09 ├── robots.txt
.rw-r--r-- 1.2k tnicollet 10 Feb  9:09 ├── service-worker.js
drwxr-xr-x    - tnicollet 10 Feb  9:09 └── static
drwxr-xr-x    - tnicollet 10 Feb  9:09    ├── css
.rw-r--r-- 1.1k tnicollet 10 Feb  9:09    │  ├── main.d1b05096.chunk.css
.rw-r--r-- 1.5k tnicollet 10 Feb  9:09    │  └── main.d1b05096.chunk.css.map
drwxr-xr-x    - tnicollet 10 Feb  9:09    └── js
.rw-r--r-- 891k tnicollet 10 Feb  9:09       ├── 2.43075c54.chunk.js
.rw-r--r-- 3.6k tnicollet 10 Feb  9:09       ├── 2.43075c54.chunk.js.LICENSE.txt
.rw-r--r-- 3.8M tnicollet 10 Feb  9:09       ├── 2.43075c54.chunk.js.map
.rw-r--r-- 2.4k tnicollet 10 Feb  9:09       ├── main.e8a8522c.chunk.js
.rw-r--r-- 9.9k tnicollet 10 Feb  9:09       ├── main.e8a8522c.chunk.js.map
.rw-r--r-- 1.5k tnicollet 10 Feb  9:09       ├── runtime-main.2ba79c4d.js
.rw-r--r-- 8.3k tnicollet 10 Feb  9:09       └── runtime-main.2ba79c4d.js.map
```

**Comment by nwmqpa** [comment]
After doing a
```
pulumi stack export
```
The output is conform to what is in my disk. Every hashes of BucketObject are the same hashes that the files on my disk

You can found the complete export [here](https://pastebin.com/759zAtc4)

**Comment by lukehoban** [comment]
> I also did a minimum repository that trigger the same problems

Thanks!  I just cloned that and things seem to work as expected for me:

```
$ npm install
$ pulumi config set aws:region us-west-2
$ pulumi up --yes
$ aws s3 ls s3://crm-webkit-bucket-0c22746
2020-02-13 10:02:19     891984 2.43075c54.chunk.js
2020-02-13 10:02:17       3559 2.43075c54.chunk.js.LICENSE.txt
2020-02-13 10:02:18    3816378 2.43075c54.chunk.js.map
2020-02-13 10:02:16       1035 asset-manifest.json
2020-02-13 10:02:18       3150 favicon.ico
2020-02-13 10:02:17       2207 index.html
2020-02-13 10:02:18       5347 logo192.png
2020-02-13 10:02:17       9664 logo512.png
2020-02-13 10:02:18       1069 main.d1b05096.chunk.css
2020-02-13 10:02:17       1514 main.d1b05096.chunk.css.map
2020-02-13 10:02:16       2441 main.e8a8522c.chunk.js
2020-02-13 10:02:17       9873 main.e8a8522c.chunk.js.map
2020-02-13 10:02:17        492 manifest.json
2020-02-13 10:02:18        657 precache-manifest.3ee4d1c5bcf3b07aaebd61b1a14e081a.js
2020-02-13 10:02:16         67 robots.txt
2020-02-13 10:02:18       1547 runtime-main.2ba79c4d.js
2020-02-13 10:02:17       8266 runtime-main.2ba79c4d.js.map
2020-02-13 10:02:16       1181 service-worker.js
```

Are you able to repro with that repo from scratch?

**Comment by nwmqpa** [comment]
Yes, I gave you the bare minimum on how to reproduce the error
However, since today, builds passes perfectly, so I can't reproduce the error anymore

**Comment by lukehoban** [comment]
> However, since today, builds passes perfectly, so I can't reproduce the error anymore

Huh - that is very surprising - but glad to hear it is working now!

**Comment by lukehoban** [comment]
I'll close this out - but if anyone is able to repro this going forward - please to drop a note here and we can re-open to continue investigating whether there is still some underlying issue here.

**Comment by nwmqpa** [comment]
Hello, so it's seems that on the minimal repository that I sent you, I can now reproduce the error.
I only tried the us-east-1 and eu-west-2 regions yet.

**Comment by ChronosWS** [comment]
I am seeing this issue with pulumi v1.12.0.  I have two stacks which are uploading the same content to different keys in S3.  Looking at the stack output, for the successful one, I see:
```
"urn": "urn:pulumi:dev::30-compute::aws:s3/bucketObject:BucketObject::Playtest",
...
"outputs": {
    "source": "C:\\Users\\CLIFF~1.HUD\\AppData\\Local\\Temp\\pulumi-asset-7320cef5ca597f0779b38ad31c2fe66d13e9583d3dd6790d06e9cfacf1b9722b",
...
}
```
and for the failing one:
```
"urn": "urn:pulumi:stage::30-compute::aws:s3/bucketObject:BucketObject::Playtest",
...
"outputs": {
    "source": {
            "4dabf18193072939515e22adb298388d": "c44067f5952c0a294b673a41bacd8c17",
            "hash": "7320cef5ca597f0779b38ad31c2fe66d13e9583d3dd6790d06e9cfacf1b9722b",
            "path": "game_configs/Playtest.config.json"
        },
...
}
```
Curiously, this source block in the failing case exactly matches the source block in the "inputs" block for this item, whereas in the successful stack, these blocks have quite different content.

Also, there are several objects which each stack is uploading which have the same content.  Only for some of these objects in the second stack does this issue occur.  Even if I delete all of the objects, refresh the stack state, and then reapply, the same objects end up with zero bytes.  I am quite mystified!

**Comment by lukehoban** [comment]
@ChronosWS Thanks for the additional detail - that does looks surprising, and like something that might possibly point at a root cause.  Re-opening to continue investigating based on this new data.

**Comment by igal1c0de4n** [comment]
+1
getting the same issue here with Pulumi 2.0
It's breaking my deployment since at least one css file has a zero size
Not sure sure how to proceed... any known workarounds?

**Comment by lukehoban** [comment]
@igal1c0de4n are you able to reproduce this in a fresh small project on your machine?  Or only inside a more complex project?  Are you able to reproduce on multiple machines?

Would love to get some sort of repro so we can investigate further.

**Comment by igal1c0de4n** [comment]
@lukehoban 
I can reproduce this issue on my local machine and on some deployment environments. 

On my local machine the Vue.js css files are consistently uploaded with zero size, while js files and other resources seem to upload with their content.

The issue reproduces consistently on two of my deployment environments (dev/qa). By sheer luck, the last production deployment successfully created the css objects on S3. 

Needless to say we'll be holding back production deployments until we figure it out. 
It'll be really helpful if we can connect and get your help on this

**Comment by igal1c0de4n** [comment]
An update might be useful for folks bumping into the same issue:

I changed the Pulumi deploy script to create the s3 bucket, but skip the part which traverses the source folder and uploads the S3 objects. 
The actual s3 upload is now done via aws cli (aws s3 cp ... --recursive) 
Although far from a perfect solution, it's an acceptable hack (for now).

Moving forward we're considering additional options, including replacing Pulumi with a different deployment method, unfortunately :(

**Comment by lukehoban** [comment]
After offline conversation with @igal1c0de4n I think we've narrowed down the issue here.  

There are some temp files used to optimize performance that are stored in `$TMPDRI/pulumi-asset-*` where the suffix is the hash of the file contents.  This is used to avoid unnecessary repetitive file I/O.  

It appears that the root cause of the issues reported here is that some of those files end up being 0 bytes.  Since they are just a performance optimization, **anyone hitting this should be able to remove those files (`rm $TMPDIR/pulumi-asset-*`) to avoid issues with 0 byte uploads**.

That does not explain why Pulumi is ending up corrupting these files in the first place.  We will look into that and get it fixed.

**Comment by igal1c0de4n** [comment]
@lukehoban Would love to know when the fix is released

**Comment by lukehoban** [comment]
For a community slack thread with some more investigation on this:

——

What if a stack simultaneously deletes file A and uploads file B, but those files have identical content?  I noticed in the logs that a file I was deleting and another I was uploading had identical hashes:

```
 + aws:s3/bucketObject:BucketObjectN: (create)
        [urn=urn:pulumi:deploy-ci-ap-1578-contacts-graphql::deploy::aws:s3/bucketObject:BucketObject::static/js/main.1f4cd4af.chunk.js-upd]
        contentType : "application/javascript"
        forceDestroy: false
        key         : "static/js/main.1f4cd4af.chunk.js"
        source      : asset(file:8ced15a) { /root/console-ci/static/js/main.1f4cd4af.chunk.js }

  - aws:s3/bucketObject:BucketObject: (delete)
      [urn=urn:pulumi:deploy-ci-ap-1578-contacts-graphql::deploy::aws:s3/bucketObject:BucketObject::static/js/main.2e52d953.chunk.js-upd]
      contentType              : "application/javascript"
      etag                     : "d41d8cd98f00b204e9800998ecf8427e"
      forceDestroy             : false
      key                      : "static/js/main.2e52d953.chunk.js"
      source                   : "/tmp/pulumi-asset-8ced15aac0a318134f59b6ac5c8144920dd0f29f94d10f7f8a7ef4579bd1300f"
```

Notice that asset(file:8ced15a) is the same prefix as the pulumi-asset file. That pulumi-asset file has 0 bytes in /tmp as well:

```
-rw------- 1 root root       0 Jun  7 02:17 /tmp/pulumi-asset-8ced15aac0a318134f59b6ac5c8144920dd0f29f94d10f7f8a7ef4579bd1300f
```

The file main.1f4cd4af.chunk.js ends up with 0 bytes in the S3 bucket (and main.2e52d953.chunk.js) is deleted. It seems that once a file is 0-bytes, it is always 0-bytes in subsequent updates as wesll.

**Comment by igal1c0de4n** [comment]
@lukehoban Is fix released in Pulumi v2.4.0?

**Comment by lukehoban** [comment]
@igal1c0de4n This library gets picked up in individual resource providers, so it depends on the version of the provider you are using, not on the version of the Pulumi CLI.  For AWS where you were hitting this, this fix was [picked up and released](https://github.com/pulumi/pulumi-aws/commit/9f1ae2ae1300a80f55886c3c808c2281446b5d5f) in the `2.9.1` release earlier this week.

**Comment by jbaileyashe** [comment]
Using pulumi 2.5.0 and this bug showed up again. Unfortunately I don't know how to reproduce, but the fix in https://github.com/pulumi/pulumi-terraform-bridge/pull/223 didn't quite do it.

**Comment by lukehoban** [comment]
> Using pulumi 2.5.0 and this bug showed up again

Note that the fix here is in `@pulumi/aws` version `2.9.1`.  The version of the `pulumi` CLI does not impact things.  I see from the snippet you shared you are using an older version of the AWS provider `2.8.0` (per `providers:aws::default_2_8_0`).

**Comment by jbaileyashe** [comment]
My bad! I'll update and if I see it again I'll report back. Thanks!
